### PR TITLE
feat(mlx-lm): add lora hypeparameters in lora layer

### PR DESCRIPTION
**Pull request by mzbac**
Add LoRA hyperparameters based on the https://github.com/microsoft/LoRA/blob/main/loralib/layers.py with slight modifications to support advanced LoRA fine-tuning.

### CODE REVIEW

File: llms/mlx_lm/tuner/lora.py [hunk 1]
```diff
@@ -58,31 +71,40 @@ def __init__(
...
     ):
         super().__init__()
 
         # Regular linear layer weights
         self.linear = nn.Linear(input_dims, output_dims, bias=bias)
 
+        if lora_dropout > 0.0:
+            self.lora_dropout = nn.Dropout(p=lora_dropout)
+        else:
+            self.lora_dropout = lambda x: x
```
Comment (awni) [inline llms/mlx_lm/tuner/lora.py:88]: Fwiw this is exactly what `Dropout.__call__` does if you give it a `0.0` value, so it's not really necessary. I don't mind if we keep it though if you like to be explicit :)

File: llms/mlx_lm/tuner/lora.py [hunk 2]
```diff
@@ -58,31 +71,40 @@ def __init__(
         self,
         input_dims: int,
         output_dims: int,
-        rank: int = 8,
+        r: int = 8,
+        lora_alpha: float = 16,
+        lora_dropout: float = 0.0,
+        scale: float = 10.0,
```
Comment (awni) [inline llms/mlx_lm/tuner/lora.py:77]: 👍 really nice to be able to configure these.

File: llms/mlx_lm/tuner/lora.py [hunk 3]
```diff
@@ -58,31 +71,40 @@ def __init__(
...
         )
-        self.lora_b = mx.zeros(shape=(rank, output_dims))
+        self.lora_b = mx.zeros(shape=(r, output_dims))
 
     def __call__(self, x):
         dtype = self.linear.weight.dtype
         if isinstance(self.linear, nn.QuantizedLinear):
             dtype = self.linear.scales.dtype
         y = self.linear(x.astype(dtype))
+        if self.training:
```
Comment (awni) [inline llms/mlx_lm/tuner/lora.py:107]: This should not be necessary, dropout checks `self.training` hence the reason to use `Module.train()` and `Module.eval()`

File: llms/mlx_lm/tuner/lora.py [hunk 4]
```diff
@@ -6,14 +6,27 @@
...
-            input_dims=input_dims, output_dims=output_dims, rank=rank, scale=scale
+            input_dims=input_dims,
+            output_dims=output_dims,
+            r=r,
+            lora_alpha=lora_alpha,
+            lora_dropout=lora_dropout,
+            scale=scale,
+            bias=bias,
         )
         lora_lin.linear = linear
```
Comment (awni) [inline llms/mlx_lm/tuner/lora.py:31]: You don't really need to pass `bias` here because it will always be overwritten by the next call to set the linear layer.

The `kwarg` is there only so people have the option to instantiate a `LoraLinear` directly (with or without a bias) rather than building from a pre-existing `Linear` layer

File: llms/mlx_lm/tuner/lora.py (hunk 1 above)
Comment (mzbac) [reply llms/mlx_lm/tuner/lora.py:88]: Does the dropout have a boundary check for p in the range of 0.0 to 1.0? I copied and pasted this from the original Lora implementation code. If mlx dropout has a boundary check for p, I don't see any reason why we would need to do this check.

**Review by awni — COMMENTED** [review]
Thanks! This looks great, really nice to take steps to configure those params.. I left a few minor comments, please check, and then we can merge 🚀

File: llms/mlx_lm/tuner/lora.py (hunk 1 above)
Comment (awni) [reply llms/mlx_lm/tuner/lora.py:88]: Indeed it does:
https://github.com/ml-explore/mlx/blob/main/python/mlx/nn/layers/dropout.py#L21-L22

**Comment by mzbac** [comment]
@awni  addressed all the comments. Thank you for helping with the review :)

**Review by awni — APPROVED** [review]
🚀

**Comment by gladjoyhub** [comment]
@mzbac
I saw the new feature of load adapter_file:
model, tokenizer = load(
    "mlx-community/phi-2-hf-4bit-mlx", adapter_file="path_to_your_adapters.npz"
)

But how to change adapter_file without reloading base model to save loading time?

**Comment by mzbac** [comment]
> @mzbac I saw the new feature of load adapter_file: model, tokenizer = load( "mlx-community/phi-2-hf-4bit-mlx", adapter_file="path_to_your_adapters.npz" )
> 
> But how to change adapter_file without reloading base model to save loading time?

Currently, we do not have the method to do that, but it should be possible. For example, before loading the Lora layers, we can check if the model already has any Lora layers. If there are Lora layers in the model, we can convert them back to linear layers and then reload the new set of Lora layers.

**Comment by gladjoyhub** [comment]
> > @mzbac I saw the new feature of load adapter_file: model, tokenizer = load( "mlx-community/phi-2-hf-4bit-mlx", adapter_file="path_to_your_adapters.npz" )
> > But how to change adapter_file without reloading base model to save loading time?
> 
> Currently, we do not have the method to do that, but it should be possible. For example, before loading the Lora layers, we can check if the model already has any Lora layers. If there are Lora layers in the model, we can convert them back to linear layers and then reload the new set of Lora layers.

Yes, like a tool replace_lora_layers(model, new_adapter_file), which uses
model.update_modules(tree_unflatten(list(linear_replacements.items())))
model.update(tree_unflatten(adapters))

But I don't know how to convert lora layers back to linear layers.

**Comment by mzbac** [comment]
> > > @mzbac I saw the new feature of load adapter_file: model, tokenizer = load( "mlx-community/phi-2-hf-4bit-mlx", adapter_file="path_to_your_adapters.npz" )
> > > But how to change adapter_file without reloading base model to save loading time?
> > 
> > 
> > Currently, we do not have the method to do that, but it should be possible. For example, before loading the Lora layers, we can check if the model already has any Lora layers. If there are Lora layers in the model, we can convert them back to linear layers and then reload the new set of Lora layers.
> 
> Yes, like a tool replace_lora_layers(model, new_adapter_file), which uses model.update_modules(tree_unflatten(list(linear_replacements.items()))) model.update(tree_unflatten(adapters))
> 
> But I don't know how to convert lora layers back to linear layers.

The Lora has original linear layer as a property, so you should be able to use that to recover the original model.

**Comment by gladjoyhub** [comment]
> > > > @mzbac I saw the new feature of load adapter_file: model, tokenizer = load( "mlx-community/phi-2-hf-4bit-mlx", adapter_file="path_to_your_adapters.npz" )
> > > > But how to change adapter_file without reloading base model to save loading time?
> > > 
> > > 
> > > Currently, we do not have the method to do that, but it should be possible. For example, before loading the Lora layers, we can check if the model already has any Lora layers. If there are Lora layers in the model, we can convert them back to linear layers and then reload the new set of Lora layers.
> > 
> > 
> > Yes, like a tool replace_lora_layers(model, new_adapter_file), which uses model.update_modules(tree_unflatten(list(linear_replacements.items()))) model.update(tree_unflatten(adapters))
> > But I don't know how to convert lora layers back to linear layers.
> 
> The Lora has original linear layer as a property, so you should be able to use that to recover the original model.

If names don't change in lora_layers, then I don't need to do again: model.update_modules(tree_unflatten(list(linear_replacements.items()))),

I can just do
model.update(tree_unflatten(adapters)) 
to change the weights for the new adapter

Is my understanding correct?
